engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/HashCodeAnyNode.java

```diff
@@ -0,0 +1,258 @@
+      throw new IllegalStateException(e);
+    }
+  }
+
+  @Specialization(guards = {
+      "interop.isDate(selfDate)",
+      "!interop.isTime(selfDate)",
+      "!interop.isTimeZone(selfDate)",
+  }, limit = "3")
+  long hashCodeForDateInterop(Object selfDate,
+      @CachedLibrary("selfDate") InteropLibrary interop) {
+    try {
+      return interop.asDate(selfDate).hashCode();
+    } catch (UnsupportedMessageException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  @Specialization(guards = {
+      "interop.isDuration(selfDuration)",
+  }, limit = "3")
+  long hashCodeForDurationInterop(Object selfDuration,
+      @CachedLibrary("selfDuration") InteropLibrary interop) {
+    try {
+      return interop.asDuration(selfDuration).hashCode();
+    } catch (UnsupportedMessageException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  @Specialization(guards = {
+      "interop.isString(selfStr)"
+  }, limit = "3")
+  long hashCodeForString(Object selfStr,
+      @CachedLibrary("selfStr") InteropLibrary interop) {
+    try {
+      return interop.asString(selfStr).hashCode();
+    } catch (UnsupportedMessageException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  @Specialization(guards = {
+      "interop.hasArrayElements(selfArray)"
+  }, limit = "3")
+  long hashCodeForArray(Object selfArray,
+      @CachedLibrary("selfArray") InteropLibrary interop,
+      @Cached HashCodeAnyNode hashCodeNode,
+      @Cached("createCountingProfile()") LoopConditionProfile loopProfile) {
+    try {
+      long arraySize = interop.getArraySize(selfArray);
+      loopProfile.profileCounted(arraySize);
+      int[] elemHashCodes = new int[(int) arraySize];
+      for (int i = 0; loopProfile.inject(i < arraySize); i++) {
+        if (interop.isArrayElementReadable(selfArray, i)) {
+          elemHashCodes[i] = (int) hashCodeNode.execute(
+              interop.readArrayElement(selfArray, i)
+          );
+        }
+      }
+      return Arrays.hashCode(elemHashCodes);
+    } catch (UnsupportedMessageException | InvalidArrayIndexException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  @Specialization(guards = {
+      "interop.hasIdentity(selfWithIdentity)"
+  }, limit = "3")
+  long hashCodeForInteropWithIdentity(
+      Object selfWithIdentity,
+      @CachedLibrary("selfWithIdentity") InteropLibrary interop,
+      @Cached("createClassProfile()") ValueProfile objectClassProfile) {
+    try {
+      return interop.identityHashCode(objectClassProfile.profile(selfWithIdentity));
+    } catch (UnsupportedMessageException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  @Specialization(guards = {
+      "interop.isNull(selfNull)"
+  }, limit = "3")
+  long hashCodeForNull(Object selfNull,
+      @CachedLibrary("selfNull") InteropLibrary interop
+  ) {
+    return 0;
+  }
+
+  @TruffleBoundary
+  @Fallback
+  long hashCodeGeneric(Object self) {
+    return self.hashCode();
```

Comment:
This may open pandora's box - what `self` do we delegate to? It may be safer to throw exception here and handle known cases by real specializations.

Reply:
Agree. Throwing away this fallback.